Add proper type annotations for RealmEmojiType and ReactionType.

roberthoenig opened the pull request:
I'm fairly confident about the first commit, but not about the second commit. There should probably be a more streamlined approach to annotating Reactions, I just couldn't find one.

borisyankov commented:
Pretty cool. One of the reasons Reactions weren't easy to type was that we had at least two (and seemingly three) related entries.

I am not sure about the naming, I would prefer `MessageReactionInfo` to `MessageReaction`, but also maybe more specific typing. 

I wonder how we differentiate these on the backend? Maybe good idea to reuse their naming.

roberthoenig commented:
Okay, so I inspected this a bit more. Basically, with some small change to the code it is possible to just dump `MessageReactionInfo` and use only to different types: `ReactionEvent` and `ReactionCounter`. `ReactionEvent` would equal `Reaction` in this PR. This reduction to only two instead of three types can be accomplished by including *all* event properties in the `EventReactionAction` in `eventToAction` and setting them in the reducer `chatReducer`.

Does this sound good to proceed?

roberthoenig commented:
@gnprice I updated both commits and added some more. For the emoji commit, I looked through the codebase and found some more places where RealmEmojiType was incorrectly used for maps. RealmEmojiState should be the correct type. 

I also rewrote the reaction type commits. They should reflect now what's going on in the current code; however, I did not yet refactor the code to simplify those types.

borisyankov commented on src/types.js:
```diff
@@ -47,13 +47,25 @@ export type InputSelectionType = {
-  name: string,
+};
+
+export type EventReaction = {
```
You can use a single type and make `emoji_core` and `reaction_type` optional.

roberthoenig replied:
The problem I experienced with that is that I get an error like this:
```.js
Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ src/reactions/aggregateReactions.js:5:3

Cannot return Array.from(...) because:
 • undefined [1] is incompatible with string [2] in property code of array element.
 • undefined [3] is incompatible with string [4] in property type of array element.

     src/reactions/aggregateReactions.js
      2│ import type { EventReaction, AggregatedReaction } from '../types';
      3│
      4│ export default (reactions: EventReaction[], ownEmail: string): AggregatedReaction[] =>
      5│   Array.from(
      6│     reactions
      7│       .reduce((reactionMap, x) => {
      8│         if (!reactionMap.has(x.emoji_name)) {
      9│           reactionMap.set(x.emoji_name, {
     10│             code: x.emoji_code,
     11│             count: 1,
     12│             name: x.emoji_name,
     13│             type: x.reaction_type,
     14│             selfReacted: false,
     15│           });
     16│         } else {
     17│           const prevReaction = reactionMap.get(x.emoji_name);
     18│           if (prevReaction) {
     19│             reactionMap.set(x.emoji_name, {
     20│               ...prevReaction,
     21│               count: prevReaction.count + 1,
     22│             });
     23│           }
     24│         }
     25│
     26│         if (x.user && x.user.email === ownEmail) {
     27│           reactionMap.set(x.emoji_name, {
     28│             ...reactionMap.get(x.emoji_name),
     29│             selfReacted: true,
     30│           });
     31│         }
     32│
     33│         return reactionMap;
     34│       }, new Map())
     35│       .values(),
     36│   );
     37│

     src/types.js
 [1] 51│   emoji_code?: string,
     52│   emoji_name: string,
 [3] 53│   reaction_type?: string,
       :
 [2] 58│   code: string,
     59│   count: number,
     60│   name: string,
     61│   selfReacted: boolean,
 [4] 62│   type: string,
```

So basically, Flow must have a guarantee that `emoji_core` and `reaction_type` are set, or it'll complain. What do you suggest?

borisyankov replied:
That is a good point. There are ways to handle this but your approach looks simplest.

gnprice commented:
Thanks @roberthoenig ! Just merged this. I especially appreciate the good explanatory commit messages.

I made a few small tweaks:
* In the comment in the first commit, I added a line saying what that item is for, in addition to the link which provides details.
* I also took the period off the end of that comment, since it messes up the URL if copy-pasted; in general I'm pretty happy to break normal punctuation rules to make it totally unambiguous what belongs in a given URL or doesn't. :-)
* In the second commit, I added a line to the commit message saying why.
* I added a commit on top which draws on the very helpful message in your last commit to start putting doc comments on our types.
